### PR TITLE
update setup and requirements

### LINKED ISSUE #230: [FEATURE REQUEST] Reduce pip dependencies to inference-time requirements.
**Is your feature request related to a problem? Please describe.**
Currently, a number of the dependencies installed by default in the library are dependencies that are not required for inference in the model (such as wandb), which means that when users pip install the library, they get both the dependencies for inference, as well as the dependencies for training. This can cause issues for downstream libraries depending on this code for inference only, as they have to deal with dependency conflicts that aren't really conflicts, and increased installation size.

**Describe the workflow you want to enable.**
Alter the pip installation so that the only required dependencies are those necessary for inference.

**Describe your proposed solution.**
Move training/dev dependencies to their own optional requirements setup, i.e.
`pip install open-flamingo[training]`, `pip install open-flamingo[dev]`, `pip install open-flamingo[all]`
(or similar naming)

**Describe alternatives you've considered**
N/A - I don't know if there's any other way to do this. I suppose another way would be to tag an 'inference-only' release, or have a separate inference-only pip package, but that seems a bit excessive.

**Additional context**
This could be a breaking change in a lot of workflows, so there's a pretty wide number of things that need to be checked - this would be very helpful for usability though, and would reduce the need for library developers to maintain their own fork just for dependency resolution purposes.

**Are you willing to help implement this feature?**
Yes, however I'm not well versed in the code, so I don't know exactly what's required for inference. I ran into this issue with wandb (which for sure is not required for inference time), but there may be other dependencies that fall into the same category.

### PR DESCRIPTION
**Pull request by loftusa**
Hey all! Making this as an introductory issue - This is to fix #230. 

I'm still pretty unfamiliar with the code base, so there's a very reasonable chance that something is in the wrong place. Just let me know any issues and I'll get them fixed.

Cheers!

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,17 +2,10 @@ einops
...
-sentencepiece==0.1.98
-pycocoevalcap
-pycocotools
-open_clip_torch>=2.16.0
+open_clip_torch>=2.16.0 
+
+# more-itertools  # not used anywhere
+# datasets  # doesn't appear to be used anywhere
+# scipy  # scipy does not appear to be used in this package
+# sentencepiece==0.1.98  # does not appear to be used
```
Comment (anas-awadalla) [inline requirements.txt:11]: I think sentencepiece is used under the hood for tokenizers. Can we add this back in?

Comment (loftusa) [reply requirements.txt]: done

Comment (DavidMChan) [reply requirements.txt]: Do we need to include scipy in the main dependencies? Also, do we need to depend on sentencepiece directly? Or can we do it indirectly through tokenizers. Ideally, if tokenizers is updated to remove sentencepiece as a dependency, we shouldn't continue including it in the distribution - right?

**Comment by DavidMChan** [comment]
Do the README/any other docs or tutorials need to be updated as a result of this?

**Comment by anas-awadalla** [comment]
> Do the README/any other docs or tutorials need to be updated as a result of this?

Yes! We need to update [environment.yml](https://github.com/mlfoundations/open_flamingo/blob/main/environment.yml) to install the training/dev requirements and add instructions in the readme on the different requirements.

File: requirements.txt (hunk 1 above)
Comment (anas-awadalla) [reply requirements.txt]: Yep we can move scipy to eval requirements. Iirc sentencepiece was not installed by default in tokenizers but maybe this is no longer the case, I need to check again?

Comment (loftusa) [reply requirements.txt]: > Yep we can move scipy to eval requirements. Iirc sentencepiece was not installed by default in tokenizers but maybe this is no longer the case, I need to check again?

It's here in `transformers` --
`"sentencepiece>=0.1.91,!=0.1.92",`
https://github.com/huggingface/transformers/blob/41c5f45bfe958be58bac6c891652f632ebac23e2/setup.py#L165

**Comment by loftusa** [comment]
> > Do the README/any other docs or tutorials need to be updated as a result of this?
> 
> Yes! We need to update [environment.yml](https://github.com/mlfoundations/open_flamingo/blob/main/environment.yml) to install the training/dev requirements and add instructions in the readme on the different requirements.

Updated!

**Review by anas-awadalla — APPROVED** [review]

**Pull request merged**